src/librbd/ImageCtx.cc

```diff
@@ -850,7 +850,11 @@ struct C_InvalidateCache : public Context {
     on_safe = util::create_async_context_callback(*this, on_safe);
 
     assert(owner_lock.is_locked());
-    if (object_cacher != NULL) {
+
+    bool journaling = false;
+    journaling = (journal != NULL && !journal->is_journal_replaying());
+
+    if (!journaling && object_cacher != NULL) {
```

Comment:
Finally did a non-cursory review.  I think we want this optimization in AioImageFlush (https://github.com/ceph/ceph/blob/master/src/librbd/AioImageRequest.cc#L452).  This is the entry-point for a user-generated flush request.  This flush method is too intertwined right now w/ other expectations that we can't remove it (e.g. if releasing the lock we need to flush out all in-flight IO).